lib/streamlit/web/server/routes.py

```diff
@@ -93,6 +95,37 @@ def set_default_headers(self):
         self.set_header("Access-Control-Allow-Origin", "*")
 
 
+class AppStaticFileHandler(AssetsFileHandler):
+    def validate_absolute_path(self, root: str, absolute_path: str) -> Optional[str]:
+        full_path = os.path.realpath(absolute_path)
+
+        if os.path.isdir(full_path):
+            # we don't want to serve directories, and serve only files
+            raise tornado.web.HTTPError(404)
+
+        if os.path.commonprefix([full_path, root]) != root:
+            # Don't allow misbehaving clients to break out of the static files directory
+            _LOGGER.warning(
+                "Serving files outside of the static directory is not supported"
+            )
+            raise tornado.web.HTTPError(404)
+
+        if (
+            os.path.exists(full_path)
+            and os.path.getsize(full_path) > file_util.MAX_APP_STATIC_FILE_SIZE
+        ):
+            raise tornado.web.HTTPError(
+                404, "TOO LARGE FILE!!!", reason="TOO LARGE FILE"
+            )
+
+        return super().validate_absolute_path(root, absolute_path)
+
+    def set_extra_headers(self, path: str) -> None:
+        if Path(path).suffix not in file_util.WHITELISTED_APP_STATIC_FILE_EXTENSIONS:
+            self.set_header("Content-Type", "text/plain")
+        self.set_header("X-Content-Type-Options", "nosniff")
```

Comment:
@kajarenc hhmm.. we should test this. it might not display the image properly are we have "nosniff" and no content-type header set. ie we will need to add a header "Content-type: image/jpeg image/png" .

Reply:
Thanks for the comment, @sfc-gh-hpathak  
Tornado itself sets headers in `tornado.web.StaticFileHandler` based on file extension using `mimetypes.guess_type` (which is part of built-in python). We inherit our `AppStaticFileHandler` from StaticFileHandler.
https://github.com/tornadoweb/tornado/blob/b726f6d66133189a1cf25a695e4e1c96d70c713b/tornado/web.py#L2700
https://github.com/tornadoweb/tornado/blob/b726f6d66133189a1cf25a695e4e1c96d70c713b/tornado/web.py#L2911

In case a file in extension in `WHITELISTED_APP_STATIC_FILE_EXTENSIONS` tornado default mechanism works and the corresponding ContentType header is set.

Reply:
That behavior tested in test_static_image_200 test:
https://github.com/kajarenc/streamlit/blob/babf369df71786d7785a7024aab6ffa057597a18/lib/tests/streamlit/web/server/routes_test.py#L218